pex/interpreter_constraints.py

```diff
@@ -138,3 +146,71 @@ def create_message(self, preamble=None):
             ),
             problems=problems,
         )
+
+
+class Lifecycle(Enum):
+    class Value(Enum.Value):
+        pass
+
+    DEV = Value("dev")
+    STABLE = Value("stable")
+    EOL = Value("eol")
+
+
+@attr.s(frozen=True)
+class PythonVersion(object):
+    lifecycle = attr.ib()  # type: Lifecycle.Value
+    major = attr.ib()  # type: int
+    minor = attr.ib()  # type: int
+    patch = attr.ib()  # type: int
+
+    def pad(self, padding):
+        # type: (int) -> PythonVersion
+        if self.lifecycle == Lifecycle.EOL:
+            return self
+        return attr.evolve(self, patch=self.patch + padding)
+
+    def iter_compatible_versions(self, specifier_sets):
+        # type: (Iterable[SpecifierSet]) -> Iterator[Tuple[int, int, int]]
+        for patch in range(self.patch + 1):
+            version = (self.major, self.minor, patch)
+            version_string = ".".join(map(str, version))
+            if not specifier_sets:
+                yield version
+            else:
+                for specifier_set in specifier_sets:
+                    if version_string in specifier_set:
+                        yield version
+                        break
+
+
+# TODO(John Sirois): Integrate a `pyenv install -l` based lint / generate script for CI / local
+# use that emits the current max patch for these versions so we automatically stay up to date
+# mod dormancy in the project.
+
+COMPATIBLE_PYTHON_VERSIONS = tuple(
+    # Each 5 units of padding costs `iter_compatible_versions` ~2 extra ms and a padding of 10 gets
+    # us through most of a brand new stable release's likely active lifecycle.
+    version.pad(10)
```

Comment:
We used to use a lower number for Pants, but @benjyw updated it to 30 with the argument it's worse to have Pex/Pants not work with some future patch release like 3.10.11 than for it to be a little slower: https://github.com/pantsbuild/pants/blob/523bdb2c2844dd56e266499796dcb712ae26c694/src/python/pants/backend/python/util_rules/interpreter_constraints.py#L38

We know from Pants that people often do use older versions for longer than at least I would have anticipated.

The EOL stuff looks great, but maybe increase this number? Or let you choose the padding per major.minor: padding for `3.7.12` of 10 is probably too high, whereas for `3.11.0` it's not enough.

Reply:
I had a similar comment on this code in #1577

Reply:
30 might have been excessive, but it depends on the perf hit. Doing this per major.minor would mitigate that.

Reply:
I've gone with 30 but left the door open to plumbing an option as a backstop which is used in tests right away.